smarts/env/wrappers/parallel_env.py

```diff
@@ -0,0 +1,358 @@
+# MIT License
+#
+# Copyright (C) 2021. Huawei Technologies Co., Ltd. All rights reserved.
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+import gym
+import multiprocessing as mp
+import numpy as np
+import sys
+import warnings
+
+from gym.vector.async_vector_env import AsyncState, AsyncVectorEnv
+from gym.error import (
+    AlreadyPendingCallError,
+    NoAsyncCallError,
+)
+from smarts.env.utils.cloud_pickle import CloudpickleWrapper
+from typing import Any, Callable, Dict, Optional, Sequence, Tuple, Union
+
+__all__ = ["ParallelEnv"]
+
+
+EnvConstructor = Callable[[str], gym.Env]
+
+
+class ParallelEnv(AsyncVectorEnv):
+    """Batch together multiple environments and step them in parallel. Each
+    environment is simulated in an external process for lock-free parallelism
+    using `multiprocessing` processes, and pipes for communication.
+
+    Note:
+        Simulation might slow down when number of parallel environments
+        requested exceed number of available CPU logical cores.
```

Comment:
Generally there is the suggestion to leave 1 CPU open for the main thread. Performance also generally improves by the number of physical cores rather than logical ones, with some exceptions.
```suggestion
        requested exceed number of available CPU logical cores -1.
```

Reply:
Dropped the terms physical and logical.